Mage/CommandCenter/CommandCenter.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Collections.Generic;
+
+using Newtonsoft.Json.Linq;
+
+
+public class CommandCenter {
+	private Mage mage { get { return Mage.Instance; }}
+	private Logger logger { get { return mage.logger("CommandCenter"); }}
+	
+	// Endpoint and credentials
+	private string baseUrl;
+	private string appName;
+	private string username;
+	private string password;
+
+	// Current transport client
+	private CommandTransportClient transportClient;
+
+	// Command Batches
+	private int nextQueryId = 1;
+	private CommandBatch currentBatch;
+	private CommandBatch sendingBatch;
+
+	//
+	public CommandCenter(CommandTransportType transportType = CommandTransportType.HTTP) {
+		currentBatch = new CommandBatch(nextQueryId++);
+		SetTransport(transportType);
+	}
+	
+	//
+	public void SetTransport(CommandTransportType transportType) {
+		// Cleanup existing transport client
+		if (transportClient != null) {
+			transportClient = null;
+		}
+
+		// Create new transport client instance
+		if (transportType == CommandTransportType.HTTP) {
+			transportClient = new CommandHTTPClient() as CommandTransportClient;
+			transportClient.SetEndpoint(baseUrl, appName, username, password);
+		} else if (transportType == CommandTransportType.JSONRPC) {
+			transportClient = new CommandJSONRPCClient() as CommandTransportClient;
+			transportClient.SetEndpoint(baseUrl, appName, username, password);
+		} else {
+			throw new Exception("Invalid transport type: " + transportType);
+		}
+
+		// Setup event handlers
+		transportClient.OnSendComplete += BatchComplete;
+		transportClient.OnTransportError += TransportError;
+	}
+
+	//
+	public void SetEndpoint(string baseUrl, string appName, string username = null, string password = null) {
+		this.baseUrl = baseUrl;
+		this.appName = appName;
+		this.username = username;
+		this.password = password;
+
+		if (transportClient != null) {
+			transportClient.SetEndpoint(baseUrl, appName, username, password);
+		}
+	}
+
+	//
+	private void SendBatch() {
+		mage.eventManager.emit("io.send", null);
+
+		lock ((object)currentBatch) {
```

Comment:
Everywhere you lock `currentBatch`, you actually use it to protect more than just `currentBatch`. In that case, why not lock `this` instead?